### PR TITLE
Improve performance with priority list

### PR DESCRIPTION
**Pull request by proteneer**

### CODE REVIEW

File: bgl/atom_mapping.py [hunk 1]
```diff
@@ -126,24 +127,45 @@ def _get_cores_impl(mol_a, mol_b, ring_cutoff, chain_cutoff, timeout, keep_conne
...
+    n_b = len(conf_b)
 
-    all_cores = mcgregor.mcs(predicate, bonds_a, bonds_b, timeout)
-    if keep_connected_component:
+    import time
+
+    start_time = time.time()
+    all_cores, timed_out = mcgregor.mcs(n_a, n_b, priority_idxs, bonds_a, bonds_b, timeout, max_cores)
+    print("mcs elapsed_time", time.time()-start_time)
+    return all_cores, timed_out
```
Comment (jkausrelay) [inline bgl/atom_mapping.py]: @proteneer remove this line

File: bgl/mcgregor.py [hunk 2]
```diff
@@ -5,9 +5,7 @@
 
 # when computed on a leaf node this is equal to the number of edges mapped.
 def arcs_left(marcs):
-    # courtesy of jfass
-    return np.sum(np.any(marcs, 1))
-
+    return np.count_nonzero(marcs)  # slow
```
Comment (maxentile) [inline bgl/mcgregor.py:8]: Observation: Initially, this was counting the number of rows containing at least one nonzero element
https://github.com/proteneer/bgl/blob/ee8d958d8a0c6862d4471d1e98242b67ae23b165/mcgregor.py#L5-L11 , but this PR will change to counting the total number of nonzero elements.

Is this change okay?

(based on p. 30 of ref, I think the initial behavior was correct, but not sure if the change makes a difference in practice)

Comment (proteneer) [reply bgl/mcgregor.py:8]: marcs used to be a list of list of bools 

now marcs is a list of integers (encoding the bit pattern of the list of bools)

Comment (maxentile) [reply bgl/mcgregor.py:8]: Ahh -- makes sense!

File: bgl/atom_mapping.py [hunk 3]
```diff
@@ -107,41 +126,116 @@ def _get_cores_impl(mol_a, mol_b, ring_cutoff, chain_cutoff, timeout, keep_conne
...
-    all_cores = mcgregor.mcs(predicate, bonds_a, bonds_b, timeout)
-    if keep_connected_component:
+    all_cores, timed_out = mcgregor.mcs(n_a, n_b, priority_idxs, bonds_a, bonds_b, timeout, max_cores)
+
+    # undo the sort
+    for core in all_cores:
+        inv_core = []
+        for atom in core[:, 0]:
+            inv_core.append(perm[atom])
+        core[:, 0] = inv_core
```
Comment (jkausrelay) [inline bgl/atom_mapping.py]: This should be done after removing disconnected components and the rmsd check, since mol_a/conf_a are still permuted here.

Comment (proteneer) [reply bgl/atom_mapping.py]: willfix

**Pull request merged**